sqs/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/sinks/SqsSink.scala

```diff
@@ -0,0 +1,335 @@
+          val response = client.sendMessageBatch(batchRequest)
+          val failures = response
+            .getFailed
+            .asScala
+            .toList
+            .map { bree =>
+              (bree.getId, BatchResultErrorInfo(bree.getCode, bree.getMessage))
+            }
+            .toMap
+          // Events to retry and reasons for failure
+          msgGroup.collect {
+            case (e, m) if (failures.keys.toSet.contains(m.getId)) =>
+              (e, failures.getOrElse(m.getId, BatchResultErrorInfo("unknown code", "unknown message")))
+          }
+        }
+        .toMap
+    }
+
+  def toSqsMessage(event: EventBuffer.Event): SendMessageBatchRequestEntry =
+    // The UUID is not used anywhere but is required by the constructor
+    new SendMessageBatchRequestEntry(UUID.randomUUID.toString, b64Encode(event))
+
+  def b64Encode(bb: ByteBuffer): String = {
+    val buffer = java.util.Base64.getEncoder.encode(bb)
+    new String(buffer.array())
+  }
+
+  def scheduleBatch(batch: List[EventBuffer.Event], lastBackoff: Long = minBackoff): Unit = {
+    val nextBackoff = getNextBackoff(lastBackoff)
+    executorService.schedule(new Thread {
+      override def run(): Unit =
+        sendBatch(batch, nextBackoff)
+    }, lastBackoff, MILLISECONDS)
+    ()
+  }
+
+  /**
+    * How long to wait before sending the next request
+    * @param lastBackoff The previous backoff time
+    * @return Minimum of maxBackoff and a random number between minBackoff and three times lastBackoff
+    */
+  private def getNextBackoff(lastBackoff: Long): Long =
+    (minBackoff + randomGenerator.nextDouble() * (lastBackoff * 3 - minBackoff)).toLong.min(maxBackoff)
+
+  def shutdown(): Unit = {
+    executorService.shutdown()
+    executorService.awaitTermination(10000, MILLISECONDS)
+    ()
+  }
+
+  case class BatchResultErrorInfo private (code: String, message: String)
+}
+
+/**
+  *  SqsSink companion object with factory method
+  */
+object SqsSink {
+
+  /**
+    * Create an SqsSink and schedule a task to flush its EventBuffer.
+    * Exists so that no threads can get a reference to the SqsSink
+    * during its construction.
+    */
+  def createAndInitialize(
+    sqsConfig: Sqs,
+    bufferConfig: BufferConfig,
+    queueName: String,
+    executorService: ScheduledExecutorService
+  ): Either[Throwable, SqsSink] = {
+    val client = for {
+      provider <- getProvider(sqsConfig.aws)
+      client   <- createSqsClient(provider, sqsConfig.region)
+      _ = sqsQueueExists(client, queueName)
+    } yield client
+
+    client.map { c =>
+      val sqsSink = new SqsSink(c, sqsConfig, bufferConfig, queueName, executorService)
+      sqsSink.EventBuffer.scheduleFlush()
+
+      // When the application is shut down try to send all buffered events.
+      Runtime
+        .getRuntime
+        .addShutdownHook(new Thread {
+          override def run(): Unit = {
+            sqsSink.EventBuffer.flush()
```

Comment:
I think that `flush()` should be inside `shutdown()` function

Reply:
I decided not to do this now (better to change it in all places) but I'm interested what benefits you see in putting it there.

Reply:
I think that there is no scenario where we would want to shutdown without flushing.